content/blog/2025-03-21-parquet-pushdown.md

```diff
@@ -0,0 +1,259 @@
+This eliminates the waste of decoding rows that will be filtered out.
+
+While simple in theory, practical implementations often make performance worse.
+
+## Why slower?
+
+At a high level, the Parquet reader first builds a filter mask -- essentially a boolean array indicating which rows meet the filter criteria -- and then uses this mask to selectively decode only the needed rows from the remaining columns in the projection.
+
+Let's dig into details of [how filter pushdown is implemented](https://github.com/apache/arrow-rs/blob/d5339f31a60a4bd8a4256e7120fe32603249d88e/parquet/src/arrow/async_reader/mod.rs#L618-L712) in the current Rust implementation of Parquet readers.
+
+<img src="/blog/images/parquet-pushdown/baseline-impl.jpg" alt="Implementation of filter pushdown in Rust Parquet readers -- the first phase builds the filter mask, the second phase applies the filter mask to the other columns" width="80%" class="img-responsive">
+
+The filter pushdown has two phases:
+
+1. Build the filter mask (steps 1-3)
+
+2. Apply the filter mask to the other columns (steps 4-7)
+
+Within each phase, it takes three steps from Parquet to Arrow:
+
+1. Decompress the Parquet pages using generic decompression algorithms like LZ4, Zstd, etc. (steps 1, 4, 6)
+
+2. Decode the page content into Arrow format (steps 2, 5, 7)
+
+3. Evaluate the filter over Arrow data (step 3)
+
+In the figure above, we can see that `location` is **decompressed and decoded twice**, first when building the filter mask (steps 1, 2), and second when building the output (steps 4, 5).
+This happens for all columns that appear both in the filter and output.
+
+The table below shows the corresponding CPU time on the [ClickBench query 22](https://github.com/apache/datafusion/blob/main/benchmarks/queries/clickbench/queries.sql#L23):
+
+```
++------------+--------+-------------+--------+
+| Decompress | Decode | Apply filter| Others |
++------------+--------+-------------+--------+
+| 206 ms     | 117 ms | 22 ms       | 48 ms  |
++------------+--------+-------------+--------+
+```
+
+Clearly, decompress/decode operations dominate the time spent. With filter pushdown, we need to decompress/decode three times; but without filter pushdown, we only need to do this twice.
+This explains why filter pushdown is slower.
+
+
+> **Note:** Highly selective filters may skip the entire page; but as long as we read one row from the page, we need to decompress/decode the entire page.
+
+
+## Attempt: cache filter columns
+
+Intuitively, caching the filter columns and reusing them later could help.
+
+But caching consumes prohibitively high memory:
+
+1. We need to cache Arrow arrays, which are on average [4x larger than Parquet data](https://github.com/XiangpengHao/liquid-cache/blob/main/dev/doc/liquid-cache-vldb.pdf).
+
+2. We need to cache the **entire column in memory**, because in Phase 1 we build filters over the entire column, and only use it in Phase 2.  
```

Comment:
I think technically it is the entire "column chunk" (aka all the values in the column for a particular row group)